core/sql/sqlcomp/CmpSeabaseDDLroutine.cpp

```diff
@@ -1925,3 +2730,568 @@ short CmpSeabaseDDL::createSeabaseLibmgrCPPLib(ExeCliInterface * cliInterface)
+
+      if ((! lti.newName) || (! lti.oldName) || (NOT lti.upgradeNeeded))
+        continue;
+
+      if (newToOld)
+        str_sprintf(queryBuf, "alter table %s.\"%s\".%s rename to %s ; ",
+                    getSystemCatalog(), SEABASE_MD_SCHEMA, lti.newName, lti.oldName);
+      else
+        str_sprintf(queryBuf, "alter table %s.\"%s\".%s rename to %s ; ",
+                    getSystemCatalog(), SEABASE_MD_SCHEMA, lti.oldName, lti.newName);
+        
+      cliRC = cliInterface->executeImmediate(queryBuf);
+      if (cliRC == -1389 || cliRC == -1390)
+        {
+          // ignore.
+          cliRC = 0;
+        }
+      else if (cliRC < 0)
+	{
+	  cliInterface->retrieveSQLDiagnostics(CmpCommon::diags());
+	}
+    }
+
+  return 0;
+}
+
+short CmpSeabaseDDL::copyOldLibrariesToNew(ExeCliInterface * cliInterface)
+{
+  Lng32 cliRC = 0;
+  NAString failedLibraries;
+
+  char queryBuf[10000];
+  for (Int32 i = 0; i < sizeof(allLibrariesUpgradeInfo)/sizeof(MDUpgradeInfo); i++)
+    {
+      const MDUpgradeInfo lti = allLibrariesUpgradeInfo[i];
+
+      if ((! lti.newName) || (! lti.oldName) || (NOT lti.upgradeNeeded))
+        continue;
+      // Update all existing libraries  so the blob contains the library
+      char * sbuf = new(STMTHEAP) char[200];
+      char * ubuf = new(STMTHEAP) char[500];
+      NABoolean libsToUpgrade = TRUE;
+      Queue *userLibsQ = NULL;
+      str_sprintf(sbuf, "select library_filename,library_uid from %s.\"%s\".%s"
+                  " for read uncommitted access",
+                  getSystemCatalog(), SEABASE_MD_SCHEMA, lti.oldName); 
+      cliRC = cliInterface->fetchAllRows(userLibsQ,sbuf, 0, FALSE,FALSE,TRUE/*no exec*/);
+      if (cliRC < 0)
+        {
+          cliInterface->retrieveSQLDiagnostics(CmpCommon::diags());
+          return -1;
+        }
+
+    
+     
+      str_sprintf(queryBuf, "upsert using load into %s.\"%s\".%s %s%s%s select %s from %s.\"%s\".%s SRC %s;",
+                  TRAFODION_SYSCAT_LIT,
+                  SEABASE_MD_SCHEMA,
+                  lti.newName, 
+                  (lti.insertedCols ? "(" : ""),
+                  (lti.insertedCols ? lti.selectedCols : ""), // insert only the original column values
+                  (lti.insertedCols ? ")" : ""),
+                  (lti.selectedCols ? lti.selectedCols : "*"),
+                  TRAFODION_SYSCAT_LIT,
+                  SEABASE_MD_SCHEMA,
+                  lti.oldName,
+                  (lti.wherePred ? lti.wherePred : ""));
+
+      cliRC = cliInterface->executeImmediate(queryBuf);
+      if (cliRC < 0)
+        {
+          cliInterface->retrieveSQLDiagnostics(CmpCommon::diags());
+          return -1;
+        }
+
+      //update contents into  the new library table so the libname doesn't contain the path and the 
+      //lib contents get loaded into the blob column.
+      userLibsQ->position();
+      for (size_t i = 0; i < userLibsQ->numEntries(); i++)
```

Comment:
This says it loops over userLibsQ, assuming that these are user libraries, but I don't see where it would exclude our system libraries for the ValidateRoutine SPJ and the predefined UDFs.

Reply:
Included the system libraries in this upgrade. So they will also have an entry in the blob volumn. There is one exception to this - validateRoutine that would show up as a "failed" upgrade but it doesn't prevent the upgrade operation.

Reply:
Thanks for the explanation.